Add docker-compose project to run all apps

LuizFDS opened the pull request:
With docker-compose we now setup it as startup project and run it from Visual Studio. It will spin up both apps todoapi and todo-web-server plus jaeger and prometheus.

![image](https://user-images.githubusercontent.com/6718493/204679596-1d38d5e2-881d-4a8c-a3fc-bdabdab4838a.png)

davidfowl commented:
I like the docker compose idea but there are a few things I don't like:
- Why is there a specific open telemetry section instead of just using the config entries for individual providers as a signal that they should be turned on?
- Why do we need 3 docker-compose.yaml, docker-compose override.yml, and docker-compose.dcproj. Can we make this work with a vanilla call to docker compose up without the VS tooling support in there?
- Why is there a new launch settings file?
- Why is `Microsoft.VisualStudio.Azure.Containers.Tools.Targets` a dependency at all? I'd like this to not be a part of the project.

Overall, I love the idea, but I'd like it to be a bit less VS specific.

LuizFDS commented:
@davidfowl, thanks for the inputs. Please see all my replies below.

> * Why is there a specific open telemetry section instead of just using the config entries for individual providers as a signal that they should be turned on?

I've fixed it —no reason for it to be there.

> * Why do we need 3 docker-compose.yaml, docker-compose override.yml, and docker-compose.dcproj. Can we make this work with a vanilla call to docker compose up without the VS tooling support in there?

Visual Studio created it. I've removed all the VS-specific ones and left a single vanilla `docker-compose.yaml`.

> * Why is there a new launch settings file?

This one was specific for the project `docker-compose.dcproj`. Visual Studio treats it like a csproj file.

> * Why is `Microsoft.VisualStudio.Azure.Containers.Tools.Targets` a dependency at all? I'd like this to not be a part of the project.

Removed. It was added by VS automatically when I added docker support.

davidfowl commented on README.md:
```diff
@@ -92,7 +95,7 @@ docker run -d -p 9090:9090 --name prometheus -v $PWD/prometheus.yml:/etc/prometh
 #### Spans
 
-1. Uncomment `.AddOtlpExporter` below `builder.Services.AddOpenTelemetryTracing`, in the `TodoApi/OpenTelemetryExtensions.cs` file
+1. Configure Env var `OTEL_EXPORTER_OTLP_ENDPOINT` with the right endpoint URL to enable `.AddOtlpExporter` below `builder.Services.AddOpenTelemetryTracing`, in the `TodoApi/OpenTelemetryExtensions.cs` file
```
`OTEL_EXPORTER_OTLP_ENDPOINT` looks like no so friendly env var for a .NET application using the config system. Is there any reason to use this specific env var or could it look like:  `OTLP__ENDPOINT`?

LuizFDS replied:
> `OTEL_EXPORTER_OTLP_ENDPOINT` looks like no so friendly env var for a .NET application using the config system. Is there any reason to use this specific env var or could it look like: `OTLP__ENDPOINT`?

I decided to use the same variable that `OpenTelemetry.Exporter.OpenTelemetryProtocol` uses to configure the endpoint internally in the library. I can change/add one specific for the todoapi if you want to.

LuizFDS replied:
Here's OpenTelemetry Protocol Exporter [doc](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#:~:text=4318%20%5B1%5D-,Env%20vars%3A%20OTEL_EXPORTER_OTLP_ENDPOINT,-OTEL_EXPORTER_OTLP_TRACES_ENDPOINT%20OTEL_EXPORTER_OTLP_METRICS_ENDPOINT%20OTEL_EXPORTER_OTLP_LOGS_ENDPOINT).

davidfowl replied:
Thanks!

davidfowl commented on README.md:
```diff
@@ -92,7 +95,7 @@ docker run -d -p 9090:9090 --name prometheus -v $PWD/prometheus.yml:/etc/prometh
 #### Spans
 
-1. Uncomment `.AddOtlpExporter` below `builder.Services.AddOpenTelemetryTracing`, in the `TodoApi/OpenTelemetryExtensions.cs` file
+1. Configure Env var `OTEL_EXPORTER_OTLP_ENDPOINT` with the right endpoint URL to enable `.AddOtlpExporter` below `builder.Services.AddOpenTelemetryTracing`, in the `TodoApi/OpenTelemetryExtensions.cs` file
```
```suggestion
1. Configure environment variable `OTEL_EXPORTER_OTLP_ENDPOINT` with the right endpoint URL to enable `.AddOtlpExporter` below `builder.Services.AddOpenTelemetryTracing`, in the `TodoApi/OpenTelemetryExtensions.cs` file
```

davidfowl commented on README.md:
```diff
@@ -104,7 +107,7 @@ docker run -d --name jaeger -e COLLECTOR_ZIPKIN_HOST_PORT=:9411 -e COLLECTOR_OTL
 #### Logs
 
-1. Uncomment `.AddOtlpExporter` below `builder.Logging.AddOpenTelemetry`, in the `TodoApi/Extensions/OpenTelemetryExtensions.cs` file
+1. Configure Env var `OTEL_EXPORTER_OTLP_ENDPOINT` with the right endpoint URL to enable `.AddOtlpExporter` below `builder.Logging.AddOpenTelemetry`, in the `TodoApi/Extensions/OpenTelemetryExtensions.cs` file
```
```suggestion
1. Configure environment variable `OTEL_EXPORTER_OTLP_ENDPOINT` with the right endpoint URL to enable `.AddOtlpExporter` below `builder.Logging.AddOpenTelemetry`, in the `TodoApi/Extensions/OpenTelemetryExtensions.cs` file
```

LuizFDS replied:
Done.

davidfowl commented on Todo.Web/Server/Todo.Web.Server.csproj:
```diff
@@ -4,6 +4,9 @@
     <Nullable>enable</Nullable>
     <ImplicitUsings>enable</ImplicitUsings>
+    <UserSecretsId>7019fce6-4c5b-45c3-9f24-5635eae3f449</UserSecretsId>
+    <DockerDefaultTargetOS>Linux</DockerDefaultTargetOS>
+    <DockerfileContext>..\..</DockerfileContext>
```
Undo?

LuizFDS replied:
Done

davidfowl commented on TodoApi/Extensions/OpenTelemetryExtensions.cs:
```diff
@@ -20,10 +21,16 @@ public static class OpenTelemetryExtensions
     public static WebApplicationBuilder AddOpenTelemetry(this WebApplicationBuilder builder)
     {
         var resourceBuilder = ResourceBuilder.CreateDefault().AddService(builder.Environment.ApplicationName);
+        var oltpEndpoint = builder.Configuration.GetValue<string>("OTEL_EXPORTER_OTLP_ENDPOINT");
```
```suggestion
        var oltpEndpoint = builder.Configuration["OTEL_EXPORTER_OTLP_ENDPOINT"];
```

LuizFDS replied:
Done

davidfowl commented on TodoApi/TodoApi.csproj:
```diff
@@ -3,6 +3,8 @@
   <PropertyGroup>
     <TargetFramework>net7.0</TargetFramework>
     <UserSecretsId>e616b306-8ad0-4843-a0e0-79b3b5655c22</UserSecretsId>
+    <DockerDefaultTargetOS>Linux</DockerDefaultTargetOS>
+    <DockerComposeProjectPath>..\docker-compose.dcproj</DockerComposeProjectPath>
```
Undo
```suggestion
```

LuizFDS replied:
Done

davidfowl commented on Todo.Web/Server/Todo.Web.Server.csproj:
```diff
@@ -4,6 +4,9 @@
     <Nullable>enable</Nullable>
     <ImplicitUsings>enable</ImplicitUsings>
+    <UserSecretsId>7019fce6-4c5b-45c3-9f24-5635eae3f449</UserSecretsId>
+    <DockerDefaultTargetOS>Linux</DockerDefaultTargetOS>
+    <DockerfileContext>..\..</DockerfileContext>
```
```suggestion
```

LuizFDS replied:
Done

davidfowl commented on docker-compose.yml:
```diff
@@ -0,0 +1,82 @@
+    volumes:
+      - ${APPDATA}/Microsoft/UserSecrets:/root/.microsoft/usersecrets:ro
+      - ${APPDATA}/ASP.NET/Https:/root/.aspnet/https:ro
+      - type: bind
+        source: ./TodoApi/Todos.db
+        target: /app/Todos.db
+      - type: bind
+        source: ./TodoApi/Todos.db-shm
+        target: /app/Todos.db-shm
+      - type: bind
+        source: ./TodoApi/Todos.db-wal
+        target: /app/Todos.db-wal
```
This seems strange, mapping the sqlite db into the container. Maybe we should move this into a folder and map the entire folder instead.

LuizFDS replied:
Yep, I found it strange too. I was trying to avoid moving it to a folder. I will do it.

LuizFDS replied:
Done. I've changed the database to a folder called `.db` and mapped it in the docker-compose file.

davidfowl commented on .dockerignore:
```diff
@@ -0,0 +1,25 @@
+**/.classpath
```
Is this still needed?

LuizFDS replied:
Do you mean the file of this line in specific? If the first one, I'd say we leave it. It will avoid copying unneeded stuff to the container.

LuizFDS replied:
Deleted in favour of using dotnet publish container.

davidfowl commented on TodoApi/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+RUN dotnet build "TodoApi.csproj" -c Release -o /app/build
+
+FROM build AS publish
+RUN dotnet publish "TodoApi.csproj" -c Release -o /app/publish /p:UseAppHost=false
```
Why `p:UseAppHost=false`?

LuizFDS replied:
I set `p:UseAppHost` to false to disable the generation of an executable file (i.e. `TodoApi.exe`) not needed on Linux.

LuizFDS replied:
Docker files were removed in favour of using dotnet publish container.

davidfowl commented:
This is looking pretty good! A few more questions!

davidfowl commented:
You know I'm going to ask you to also do the same for tye.yml when you're done 😉

cwe1ss commented:
Does anyone know if it's possible to use Docker compose with [the new .NET SDK integration for building docker images without a Dockerfile](https://learn.microsoft.com/en-us/dotnet/core/docker/publish-as-container)?

If this were possible with Docker Compose, it would maybe be great to use it here since it's so much simpler if you don't have to keep Dockerfiles up to date.

LuizFDS commented:
> 'm going to ask y

It'll be my pleasure!

LuizFDS commented:
> Does anyone know if it's possible to use Docker compose with [the new .NET SDK integration for building docker images without a Dockerfile](https://learn.microsoft.com/en-us/dotnet/core/docker/publish-as-container)?

Yes, it's possible. You will need to run `dotnet publish --os linux --arch x64 /t:PublishContainer -c Release` to build your local image first, then you can run your docker-compose file.

I might add this change to this PR.

LuizFDS commented:
@cwe1ss @davidfowl I've removed the docker files and updated the docker-compose to run both images created via dotnet publish.

LuizFDS commented on README.md:
````diff
@@ -49,7 +51,20 @@ To run the application, run both the [Todo.Web/Server](Todo.Web/Server) and [Tod
+
+   1. Build a docker image for the `Todo.Web.Server` directly from dotnet publish.
+      ```
+      dotnet publish ./Todo.Web/Server/Todo.Web.Server.csproj --os linux --arch x64 /t:PublishContainer -c Release --self-contained true
````
The flag `--self-contained true` is needed here due to this open issue [Publishing hosted Blazor WASM project breaks if not self-contained](https://github.com/dotnet/runtime/issues/75847).

LuizFDS commented on Todo.Web/Server/Todo.Web.Server.csproj:
```diff
@@ -4,10 +4,14 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.AspNetCore.Components.WebAssembly.Server" Version="7.0.0" />
+    <PackageReference Include="Microsoft.NET.Build.Containers" Version="0.2.7" />
```
This package is currently required to publish an app as a container according with this [doc](https://learn.microsoft.com/en-us/dotnet/core/docker/publish-as-container).

LuizFDS commented on TodoApi/TodoApi.csproj:
```diff
@@ -11,6 +13,7 @@
     <PackageReference Include="Microsoft.AspNetCore.OpenApi" Version="7.0.0" />
     <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="7.0.0" PrivateAssets="all" />
     <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="7.0.0" />
+    <PackageReference Include="Microsoft.NET.Build.Containers" Version="0.2.7" />
```
This package is currently required to publish an app as a container according to this [doc](https://learn.microsoft.com/en-us/dotnet/core/docker/publish-as-container).

cwe1ss commented on docker-compose.yml:
```diff
@@ -0,0 +1,65 @@
+    file: ./TodoApi/prometheus.yml
+
+services:
+  todoapi:
```
Should this service and the container image be called "todo-api" so that it's consistent with "todo-web-server"?

LuizFDS replied:
Yep. It's renamed.

davidfowl commented on docker-compose.yml:
```diff
@@ -0,0 +1,65 @@
+      - "5000:80"
+      - "5001:443"
+    volumes:
+      - ${APPDATA}/Microsoft/UserSecrets:/root/.microsoft/usersecrets:ro
+      - ${APPDATA}/ASP.NET/Https:/root/.aspnet/https:ro
```
This is windows specific.

davidfowl replied:
What happens when this runs on linux? I guess the secrets dont get mapped

LuizFDS replied:
Yeah. It doesn't get mapped. I will update and push it again with a more generic approach.

LuizFDS replied:
I end up creating two volumes map. One for Windows and the other for macOS and Linux.

LuizFDS commented:
@cwe1ss @davidfowl I finished all the pending changes. Please let me know if you are okay with the PR as it is now.

davidfowl commented:
Amazing changes. Thanks @LuizFDS !
